rclc/include/rclc/init.h

```diff
@@ -104,6 +104,112 @@ rcl_ret_t
 rclc_support_fini(
   rclc_support_t * support);
 
+/**
+ *  Allocates the rclc_support_t object on the heap.
+ *  Can be used as if no stack allocation can or should be used.
```

Comment:
as an example,  i am posting the comment here.

Why not passing `rcl_allocator_t` in the argument then use it for the allocation for `rclc_support_t`?
So that it can be any allocator specified by the user application, it does not always have to be the heap memory, could be global memory range in case of the firmware?

i believe that is more consistent design with other APIs?

Reply:
Agreed, Considering that  user wants to use custom memory allocating interfaces, passing `rcl_allocator_t` is a better idea.

Reply:
I might not fully get the idea, but I am also not familiar with rcl_allocator_t. However isn't the rcl_allocator_t proposal a chicken egg problem? Because if a language does not parse the headers, then there never is an rcl_allocator_t to allocate anything because the user does not know sizeof(rcl_allocator_t)?

Or do you propose, that the `rclc_allocator_alloc_default()` is implemented as proposed in this pull request and after having the allocator, it is handed over to the different functions which use it to allocate the other structs? That way we'd have only one explicit malloc and the rest would be allocated with rcl_allocator_t.

To get rid of this last explicit malloc, we could either provide `sizeof(rcl_allocator_t)` as a symbol or having a function `rclc_get_default_allocator` which returns an (rcl_allocator_t *) and uses rcl_allocator_t to allocate memory for the pointed struct.

Reply:
I think these code is a demo usage of rcl_allocator_t 

https://github.com/ros2/rclc/blob/995cda8c68456c7477ed103d9419a161fd068960/rclc_examples/src/example_sub_context.c#L65
https://github.com/ros2/rclc/blob/995cda8c68456c7477ed103d9419a161fd068960/rclc_examples/src/example_sub_context.c#L114

Rather than using 'malloc'/'realloc', I believe that using allocator.allocate()/allocator.reallocate() is more scalable.
I think scalability is what @fujitatomoya  emphasized.

Reply:
I extended the pull request by adding support for the user-provided allocator.